frontend/src/test/java/com/google/sps/InMemoryPlaceVisitStorageTest.java

```diff
@@ -0,0 +1,348 @@
+                                                     .setTripId("b")
+                                                     .setUserMark("if-time")
+                                                     .setLatitude(0)
+                                                     .setLongitude(0)
+                                                     .build();
+
+  /**
+   * test that a PlaceVisitAlreadyExistsException is thrown when a place is
+   * added but another place with the same placeId and tripId has already been
+   * added to the storage
+   */
+  @Test
+  public void addPlaceVisit_samePlaceIdAndTripIdAlreadyAdded_throwsException()
+      throws PlaceVisitAlreadyExistsException {
+    InMemoryPlaceVisitStorage storage = new InMemoryPlaceVisitStorage();
+    storage.addPlaceVisit(SEOUL);
+    Assert.assertThrows(
+        PlaceVisitAlreadyExistsException.class, () -> storage.addPlaceVisit(SEOUL_B));
+  }
+
+  /**
+   * test that a PlaceNotFoundException is thrown when getting a place from
+   * an empty storage
+   */
+  @Test
+  public void getPlaceVisit_emptyStorage_returnsEmptyOptional() {
+    InMemoryPlaceVisitStorage storage = new InMemoryPlaceVisitStorage();
+    assertThat(storage.getPlaceVisit(TOKYO.tripId(), TOKYO.placeId())).isEmpty();
+  }
+
+  /**
+   * test that a PlaceNotFoundException is thrown when trying to get a place
+   * from storage when there is no place in storage with the given tripId and
+   * placeId
+   */
+  @Test
+  public void getPlaceVisit_placeNotInStorage_returnsEmptyOptional()
+      throws PlaceVisitAlreadyExistsException, PlaceVisitNotFoundException {
+    InMemoryPlaceVisitStorage storage = new InMemoryPlaceVisitStorage();
+    storage.addPlaceVisit(TOKYO);
+    storage.addPlaceVisit(TOKYO_B);
+    assertThat(storage.getPlaceVisit(SEOUL.tripId(), SEOUL.placeId())).isEmpty();
+  }
+
+  /**
+   * test getPlaceVisit when PlaceVisitModel object with given tripId and
+   * placeId exist in storage
+   */
+  @Test
+  public void addPlaceVisitGetPlaceVisit_placeAddedEqualsPlaceGot()
+      throws PlaceVisitAlreadyExistsException, PlaceVisitNotFoundException {
+    InMemoryPlaceVisitStorage storage = new InMemoryPlaceVisitStorage();
+    storage.addPlaceVisit(TOKYO);
+    storage.addPlaceVisit(TOKYO_B);
+
+    Optional<PlaceVisitModel> tokyo = storage.getPlaceVisit(TOKYO.tripId(), TOKYO.placeId());
+    assertThat(tokyo).hasValue(TOKYO);
+
+    Optional<PlaceVisitModel> tokyoB = storage.getPlaceVisit(TOKYO_B.tripId(), TOKYO_B.placeId());
+    assertThat(tokyoB).hasValue(TOKYO_B);
+  }
+
+  /**
+   * test that a PlaceNotFoundException is thrown when removing a place from
+   * an empty storage
+   */
+  @Test
+  public void removePlaceVisit_emptyStorage_throwsException() throws PlaceVisitNotFoundException {
+    InMemoryPlaceVisitStorage storage = new InMemoryPlaceVisitStorage();
+    Assert.assertThrows(PlaceVisitNotFoundException.class,
```

Comment:
FYI, you can also use a static import of `assertThrows`. Up to you if that's worth it.